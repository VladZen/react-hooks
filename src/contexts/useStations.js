import { useState, useCallback } from 'react'
import constate from 'constate'

import Api from '@/api/Stations'

const useStations = () => {
  const [stationsAreLoading, setLoading] = useState(false)
  const [stations, setStations] = useState([])

  const fetchStations = useCallback(async () => {
    setLoading(true)
    try {
      const { data: stations } = await Api.getList()
      setStations(stations)
    } catch (e) {
      console.log(e)
    } finally {
      setLoading(false)
    }
  }, [])

  const getStation = async (id) => {
    if (!stations.length) await fetchStations()
    return stations.find(({ station_ID }) => station_ID === id)
  }

  return {
    stations,
    getStation,
    fetchStations,
    stationsAreLoading
  }
}

// 2️⃣ Wrap your hook with the constate factory
const [StationsProvider, useStationsContext] = constate(useStations)

export { StationsProvider, useStationsContext }
