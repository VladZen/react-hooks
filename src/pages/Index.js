import { useStationsContext } from '@/contexts/useStations'

import Header from '@/components/Header'
import StationsList from '@/components/stations-list/StationsList'
import Loading from '@/components/Loading'

const Index = () => {
  const {
    isLoading,
    stations
  } = useStationsContext()

  return (
    <div className='layout'>
      <Header heading='Your stations' />

      <main className='layout__main'>
        <Loading trigger={isLoading}>
          <StationsList stations={ stations } />
        </Loading>
      </main>
    </div>
  )
}

export default Index
