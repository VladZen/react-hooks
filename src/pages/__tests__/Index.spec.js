import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { useStationsContext } from '@/contexts/useStations'
import Index from '@/pages/Index'

jest.mock('@/contexts/useStations', () => ({
  ...jest.requireActual('@/contexts/useStations'), // use actual for all non-hook parts
  useStationsContext: jest.fn()
}))

const stations = [
  {
    'station_ID': 103,
    'custom_evse_id': null,
    'location_ID': 103,
    'seller_ID': 27,
    'name': 'Endesa 3',
    'connected': 1,
    'position': '60.868623,26.702901',
    'available': 0,
    'lastconnect': '2019-03-26T06:46:32.000Z',
    'roaming_identifier_cpo': 'FI*001'
  },
  {
    'station_ID': 104,
    'custom_evse_id': null,
    'location_ID': 104,
    'seller_ID': 27,
    'name': 'Endesa 4',
    'connected': 1,
    'position': '60.868623,26.702901',
    'available': 1,
    'lastconnect': '2018-03-15T11:25:47.000Z',
    'roaming_identifier_cpo': 'FI*001'
  }
]

describe('Index page ', () => {
  describe('rendering', () => {
    it('loading', () => {
      useStationsContext.mockReturnValueOnce({
        isLoading: true,
        stations: []
      })

      const { container } = render(
        <Index />
      )

      expect(container).toMatchSnapshot()
    })

    it('loaded', () => {
      useStationsContext.mockReturnValueOnce({
        isLoading: false,
        stations
      })

      const { container } = render(
        <MemoryRouter>
          <Index />
        </MemoryRouter>
      )

      expect(container).toMatchSnapshot()
    })
  })
})

