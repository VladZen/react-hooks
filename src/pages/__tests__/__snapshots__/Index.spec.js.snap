// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page  rendering loaded 1`] = `
<div>
  <div
    class="layout"
  >
    <header
      class="layout__header"
    >
      
      <h1
        class="text-ellipsis"
        title="Your stations"
      >
        Your stations
      </h1>
    </header>
    <main
      class="layout__main"
    >
      <ul
        class="stations"
      >
        <li
          class="stations__item"
        >
          <a
            class="stations__item__name text-ellipsis"
            href="/stations/103"
            title="Endesa 3"
          >
            Endesa 3
          </a>
          <span
            class="stations__item__status stations__item__status--offline"
          >
            <svg
              class="stations__item__status__icon"
            >
              <use
                href="sprite.svg#offline"
              />
            </svg>
            <span
              class="text-capitalize"
            >
              offline
            </span>
          </span>
        </li>
        <li
          class="stations__item"
        >
          <a
            class="stations__item__name text-ellipsis"
            href="/stations/104"
            title="Endesa 4"
          >
            Endesa 4
          </a>
          <span
            class="stations__item__status stations__item__status--offline"
          >
            <svg
              class="stations__item__status__icon"
            >
              <use
                href="sprite.svg#available"
              />
            </svg>
            <span
              class="text-capitalize"
            >
              available
            </span>
          </span>
        </li>
      </ul>
    </main>
  </div>
</div>
`;

exports[`Index page  rendering loading 1`] = `
<div>
  <div
    class="layout"
  >
    <header
      class="layout__header"
    >
      
      <h1
        class="text-ellipsis"
        title="Your stations"
      >
        Your stations
      </h1>
    </header>
    <main
      class="layout__main"
    >
      <div
        class="spinner"
      >
        Loading...
      </div>
    </main>
  </div>
</div>
`;
